### PR TITLE
Add doc about Java/RxJava support

### PR DESCRIPTION
**Pull request by BoD**
(part of #3490)

### CODE REVIEW

File: docs/source/advanced/java-rxjava.mdx [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+Because Apollo Android is written in Kotlin, it can be used in Java projects. This guide will show you how to do that.
+
+## Use the Java codegen
+
+Apollo Android generates Kotlin code by default, but you can configure it to use the Java codegen instead:
+
+```kotlin:title=build.gradle[.kts]
+apollo {
+  generateKotlinModels.set(false)
```
Comment (martinbonnin) [inline docs/source/advanced/java-rxjava.mdx:16]: Feels weird to generate Java models but have samples in Kotlin below. What about splitting in 2 pages?

* java (contains only Java codeblocks): explains how to generate Java models and build an ApolloClient in Java and links to the rxjava page 
* rxjava (contains Java + Kotlin codeblocks): explains how to use RxJava wrappers in both Java and Kotlin

File: docs/source/advanced/java-rxjava.mdx [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+}
+```
+
+## Use RxJava adapters
```
Comment (martinbonnin) [inline docs/source/advanced/java-rxjava.mdx:20]: I'd try to stay clear of "adapters" here as it's pretty loaded already. Maybe "RxJava extensions"?

File: docs/source/advanced/java-rxjava.mdx [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    ): Flowable<ApolloResponse<D>> {
+        val modifiedRequest = request.newBuilder() /* ... */ .build()
+        return chain.proceed(modifiedRequest)
+    }
+}
+```
+</MultiCodeBlock>
+
+
+Then use the `toApolloInterceptor` extension to convert it to an `ApolloInterceptor` and add it to your `ApolloClient.Builder`:
```
Comment (martinbonnin) [inline docs/source/advanced/java-rxjava.mdx:113]: Use the fully qualified class name here ? If I'm not mistaken, something like `ToCoroutinesKt. toApolloInterceptor`?

(That doesn't look too good TBH, certainly something to fine tune)

File: docs/source/advanced/java-rxjava.mdx [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+mutationResponse.subscribe(/* ... */)
+
+// A subscription
+val subscriptionCall = client.subscribe(MySubscription())
+val subscriptionResponse: Flowable<ApolloResponse<MySubscription.Data>> = subscriptionCall.rxFlowable()
+subscriptionResponse.subscribe(/* ... */)
+```
+</MultiCodeBlock>
+
+### ApolloInterceptor
```
Comment (martinbonnin) [inline docs/source/advanced/java-rxjava.mdx:79]: For this one an the ones below, I'd mark them as `Experimental` for now as the ergonomics could certainly be improved and we've had limited feedback about them so keeping the ability to change them would be nice.

File: docs/source/advanced/java-rxjava.mdx (hunk 2 above)
Comment (BoD) [reply docs/source/advanced/java-rxjava.mdx:20]: 👍 agreed

File: docs/source/advanced/java-rxjava.mdx (hunk 4 above)
Comment (BoD) [reply docs/source/advanced/java-rxjava.mdx:79]: 👍  I will add a sentence here.

OK for me to also create/add the annotation on this PR, or preferable to make another PR for that?

File: docs/source/advanced/java-rxjava.mdx (hunk 3 above)
Comment (BoD) [reply docs/source/advanced/java-rxjava.mdx:113]: Oh yeah I can explicitly show the import and yeah importing the function also works in Java but I agree it's just clearer if we show it here.

File: docs/source/advanced/java-rxjava.mdx (hunk 1 above)
Comment (BoD) [reply docs/source/advanced/java-rxjava.mdx:16]: That sounds good 👍  Will do.

File: docs/source/advanced/java-rxjava.mdx (hunk 4 above)
Comment (martinbonnin) [reply docs/source/advanced/java-rxjava.mdx:79]: > OK for me to also create/add the annotation on this PR, or preferable to make another PR for that?
I don't know 😅 . This PR is "only" 215 lines so I guess it's ok to beef it up?

File: apollo-api/api/apollo-api.api [hunk 5]
```diff
@@ -37,6 +37,9 @@ public final class com/apollographql/apollo3/api/AnyAdapter : com/apollographql/
 	public final fun toJson (Lcom/apollographql/apollo3/api/json/JsonWriter;Ljava/lang/Object;)V
 }
 
+public abstract interface annotation class com/apollographql/apollo3/api/ApolloExperimental : java/lang/annotation/Annotation {
```
Comment (martinbonnin) [inline apollo-api/api/apollo-api.api:40]: 👍

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloExperimental.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.apollographql.apollo3.api
+
+@RequiresOptIn(
+    level = RequiresOptIn.Level.WARNING,
+    message = "This API is experimental and subject to change, use with caution"
```
Comment (martinbonnin) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloExperimental.kt:5]: Nit: "use with caution" might scare legitimate users? I'd keep it factual:

```suggestion
    message = "This API is experimental and can be changed in a backwards-incompatible manner."
```

File: apollo-rx2-support/src/main/java/com/apollographql/apollo3/rx2/toRx2.kt [hunk 7]
```diff
@@ -67,14 +68,17 @@ fun ApolloInterceptor.toRx2ApolloInterceptor(scheduler: Scheduler = Schedulers.i
...
 fun ApolloStore.toRx2ApolloStore(scheduler: Scheduler = Schedulers.io()) = Rx2ApolloStore(this, scheduler)
 
+@ApolloExperimental
 class Rx2ApolloStore(
     private val delegate: ApolloStore,
     private val scheduler: Scheduler,
 ) {
   private val dispatcher = scheduler.asCoroutineDispatcher()
 
+  @ApolloExperimental
```
Comment (martinbonnin) [inline apollo-rx2-support/src/main/java/com/apollographql/apollo3/rx2/toRx2.kt:81]: Do we need to put `@ApolloExperimental` on individual methods if the enclosing class is also `@ApolloExperimental`

File: build.gradle.kts [hunk 8]
```diff
@@ -197,4 +197,10 @@ configure<kotlinx.validation.ApiValidationExtension> {
           "apollo-testing-support"
       )
   )
-}
\ No newline at end of file
+  nonPublicMarkers.addAll(
```
Comment (martinbonnin) [inline build.gradle.kts:200]: That's cool ✨ !

File: docs/source/advanced/rxjava.mdx [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+ApolloSubscriptionCall<MySubscription.Data> subscriptionCall = client.subscribe(new MySubscription());
+Flowable<ApolloResponse<MySubscription.Data>> subscriptionResponse = Rx2Apollo.rxFlowable(subscriptionCall);
+subscriptionResponse.subscribe( /* ... */ );
+```
+
+```kotlin
+// A query
+val queryCall: ApolloQueryCall<MyQuery.Data> = client.query(MyQuery())
+val queryResponse: Single<ApolloResponse<MyQuery.Data>> = queryCall.rxSingle()
+queryResponse.subscribe(/* ... */)
```
Comment (martinbonnin) [inline docs/source/advanced/rxjava.mdx:55]:
```suggestion
client.query(MyQuery()).rxSingle().subscribe(/* ... */)
```

File: docs/source/advanced/rxjava.mdx [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+// A mutation
+val mutationCall = client.mutate(MyMutation("my-parameter"))
+val mutationResponse: Single<ApolloResponse<MyMutation.Data>> = mutationCall.rxSingle()
+mutationResponse.subscribe(/* ... */)
```
Comment (martinbonnin) [inline docs/source/advanced/rxjava.mdx:60]:
```suggestion
client.mutate(MyMutation("my-parameter")).rxSingle().subscribe(/* ... */)
```

File: docs/source/advanced/rxjava.mdx [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+// A subscription
+val subscriptionCall = client.subscribe(MySubscription())
+val subscriptionResponse: Flowable<ApolloResponse<MySubscription.Data>> = subscriptionCall.rxFlowable()
+subscriptionResponse.subscribe(/* ... */)
```
Comment (martinbonnin) [inline docs/source/advanced/rxjava.mdx:65]:
```suggestion
client.subscribe(MySubscription()).rxFlowable().subscribe(/* ... */)
```

File: apollo-rx2-support/src/main/java/com/apollographql/apollo3/rx2/toRx2.kt (hunk 7 above)
Comment (BoD) [reply apollo-rx2-support/src/main/java/com/apollographql/apollo3/rx2/toRx2.kt:81]: I was not 100% sure, and after checking, it looks like it's not useful: the method calls are marked with the warning even if they are not annotated - so I've removed them 👍 

Screen Shot 2021-11-08 at 09 40 24

**Review by martinbonnin — APPROVED** [review]

**Pull request merged**